bin/build_docker_images.sh

```diff
@@ -10,39 +18,102 @@ unset CDPATH
-
-# Tag and push the latest version of the Clipper Docker images to the container registry
-# for the Kubernetes testing cluster
-docker tag clipper/query_frontend:$tag 568959175238.dkr.ecr.us-west-1.amazonaws.com/clipper/query_frontend:$tag
-docker push 568959175238.dkr.ecr.us-west-1.amazonaws.com/clipper/query_frontend:$tag
-docker tag clipper/management_frontend:$tag 568959175238.dkr.ecr.us-west-1.amazonaws.com/clipper/management_frontend:$tag
-docker push 568959175238.dkr.ecr.us-west-1.amazonaws.com/clipper/management_frontend:$tag
-cd -
-
-# Build Spark JVM Container
-cd $DIR/../containers/jvm
-time docker build -t clipper/spark-scala-container:$tag -f SparkScalaContainerDockerfile ./
-cd -
-
-# Build the Python model containers
-cd $DIR/..
-
-# first build base image
-time docker build -t clipper/py-rpc:$tag -f ./RPCDockerfile ./
-time docker build --build-arg CODE_VERSION=$tag -t clipper/sum-container:$tag -f ./SumDockerfile ./
-time docker build --build-arg CODE_VERSION=$tag -t clipper/noop-container:$tag -f ./NoopDockerfile ./
-time docker build --build-arg CODE_VERSION=$tag -t clipper/python-closure-container:$tag -f ./PyClosureContainerDockerfile ./
-time docker build --build-arg CODE_VERSION=$tag -t clipper/pyspark-container:$tag -f ./PySparkContainerDockerfile ./
-time docker build --build-arg CODE_VERSION=$tag -t clipper/tf_cifar_container:$tag -f ./TensorFlowCifarDockerfile ./
-cd -
+CLIPPER_ROOT=$DIR/..
+cd $CLIPPER_ROOT
+
+# Initialize tags
+version_tag=$(<VERSION.txt)
+sha_tag=`git rev-parse --verify --short HEAD`
+
+namespace="clipper"
+
+# We build images with the SHA tag to try to prevent clobbering other images
+# being built from different branches on the same machine. This is particularly
+# useful for running these scripts on the Jenkins build cluster.
+create_image () {
+    local image=$1          # Param: The name of the image to build
+
+    local dockerfile=$2     # the Dockerfile name within the
+                            # <clipper_root>/dockerfiles directory
+
+    local public=$3        # Push the built images to Docker Hub under
+                            # the clipper namespace. Must have credentials.
+                     
+    echo "Building $namespace/$image:$sha_tag from file $dockerfile"
+    time docker build --build-arg CODE_VERSION=$sha_tag -t $namespace/$image:$sha_tag \
+        -f dockerfiles/$dockerfile $CLIPPER_ROOT
+    docker tag $namespace/$image:$sha_tag $namespace/$image:$version_tag
+
+    if [ "$publish" = true ] && [ "$public" = true ] ; then
+        echo "Publishing $namespace/$image:$sha_tag"
+        docker push $namespace/$image:$sha_tag
+        echo "Publishing $namespace/$image:$version_tag"
+        docker push $namespace/$image:$version_tag
+    fi
+}
+
+
+# Build the Clipper Docker images.
+build_images () {
+    # True and false indicate whether or not to public the image. We public public images.
```

Comment:
Nit: Can we use an alternative to the verb usage of `public`? "make the image public" is easier to understand.

Reply:
Oops. That's the result of a botched find/replace. Should read "True and false indicate whether or not to *publish* the image. We *publish* public images."